flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/BigQuerySchemaProvider.java

```diff
@@ -0,0 +1,379 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.flink.bigquery.sink.serializer;
+
+import com.google.api.client.util.Preconditions;
+import com.google.api.services.bigquery.model.TableSchema;
+import com.google.cloud.flink.bigquery.common.config.BigQueryConnectOptions;
+import com.google.cloud.flink.bigquery.common.exceptions.BigQueryConnectorException;
+import com.google.cloud.flink.bigquery.common.utils.SchemaTransform;
+import com.google.cloud.flink.bigquery.services.BigQueryServices.QueryDataClient;
+import com.google.cloud.flink.bigquery.services.BigQueryServicesFactory;
+import com.google.cloud.flink.bigquery.services.BigQueryUtils;
+import com.google.protobuf.DescriptorProtos.DescriptorProto;
+import com.google.protobuf.DescriptorProtos.FieldDescriptorProto;
+import com.google.protobuf.DescriptorProtos.FileDescriptorProto;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.DescriptorValidationException;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import org.apache.avro.LogicalType;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Class to obtain Descriptor ({@link Descriptor}) required for serialization of Generic Records.
+ * This class has essential methods to obtain bigQuery Table Schema ({@link TableSchema}) from the
+ * provided Table Connection Options ({@link BigQueryConnectOptions}). It is also responsible for
+ * converting this Table Schema to Avro Schema ({@link Schema}) which is used to further convert to
+ * DescriptorProto({@link DescriptorProto}). The obtained Descriptor Proto is required for the
+ * formation of the Descriptors which is essential for Proto Rows formation.
+ */
+public class BigQuerySchemaProvider implements Serializable {
+
+    private Schema avroSchema;
+    private DescriptorProto descriptorProto;
+
+    private static final Map<Schema.Type, FieldDescriptorProto.Type> AVRO_TYPES_TO_PROTO;
+    private static final Map<String, FieldDescriptorProto.Type> LOGICAL_AVRO_TYPES_TO_PROTO;
+
+    public DescriptorProto getDescriptorProto() {
+        return descriptorProto;
+    }
+
+    public Descriptor getDescriptor() {
+        Preconditions.checkNotNull(
+                getDescriptorProto(),
+                "DescriptorProto not initialized before obtaining Descriptor!");
```

Comment:
This should never happen. All constructors should derive the avro schema and desciptor proto, failing if unable to.